packages/side-nav/src/SideNav.tsx

```diff
@@ -35,11 +157,122 @@ interface SideNavProps {
+  ...rest
+}: SideNavProps) {
+  const { Provider: ContextProvider } = SideNavContext;
+  const [collapsed, setCollapsed] = useState(false);
+  const [hover, setHover] = useState(false);
+  const [focus, setFocus] = useState(false);
+  const { usingKeyboard } = useUsingKeyboardContext();
+  const navId = useMemo(() => idProp ?? navIdAllocator.generate(), [idProp]);
+  const [
+    portalContainer,
+    setPortalContainer,
+  ] = useState<HTMLUListElement | null>(null);
+
+  // We visually expand the navigation when a user focuses on an element within the navigation
+  // while navigating via keyboard.
+  const focusExpand = usingKeyboard && focus;
+
+  // Nav element should have a programmatically-determinable label
+  validateAriaLabelProps(rest, 'SideNav');
+
+  // Global event listener for toggling the navigation.
+  useEventListener(
+    'keypress',
+    e => {
+      const disabledTagNames = ['INPUT', 'TEXTAREA'] as const;
+
+      // Disable toggling the side navigation when a user is typing in an input.
+      // The typing for useEventListener doesn't seem to like using event.target,
+      // so we disable this here.
+      // @ts-expect-error
+      const shouldToggle = disabledTagNames.includes(e.target?.tagName);
+
+      if (e.keyCode === keyMap.BracketLeft && !shouldToggle) {
+        setCollapsed(curr => !curr);
+      }
+    },
+    {
+      options: {
+        passive: true,
+      },
+    },
+  );
+
   return (
-    <nav className={cx(navStyles, className)} aria-label="side-nav" {...rest}>
-      <ul className={ulStyleOverrides}>{children}</ul>
-    </nav>
+    <Transition
+      in={collapsed && !hover && !focusExpand}
+      timeout={collapseDuration}
+    >
+      {state => (
+        <ContextProvider
+          value={{
+            navId,
+            currentPath,
+            collapsed,
+            setCollapsed,
+            portalContainer: portalContainer,
+            transitionState: state,
+          }}
+        >
+          <div
+            className={cx(space, { [collapsedSpace]: collapsed }, className)}
+          >
+            <div className={wrapper} onMouseLeave={() => setHover(false)}>
+              <nav
+                id={navId}
+                className={cx(navStyles, {
+                  [collapsedNavStyles]: ['entering', 'entered'].includes(state),
+                  [hoverNavStyles]: (hover || focusExpand) && collapsed,
+                })}
+                onFocus={() => setFocus(true)}
+                onBlur={() => setFocus(false)}
+                onMouseEnter={() => setHover(true)}
+                {...rest}
+              >
+                <ul
+                  className={cx(
+                    ulStyleOverrides,
+                    listStyles,
+                    expandedListStyle,
+                    expandedStateStyles[state],
+                  )}
+                  role="menu"
+                >
+                  {children}
+                </ul>
+
+                <ul
```

Comment:
can we have a self-closing ul?

Reply:
Sorry, I meant more from an accessibility perspective. Why would we have a ul that doesn't contain any li elements.

Reply:
We don't. The elements are portaled into the list.

Reply:
See `CollapsedSideNavItem` for an example. It's how we render the collapsed state.

Reply:
Ah got it